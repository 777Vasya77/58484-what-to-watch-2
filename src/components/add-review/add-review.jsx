import React, {Fragment} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {selectors} from '~/selectors/selectors';
import PageHeader from '~/components/page-header/page-header';
import {operations} from '~/operations/oparations';

const STARS_COUNT = 5;
const validateRule = {
  TEXT: {
    MIN: 50,
    MAX: 400,
  },
  RANKING: {
    MIN: 1,
    MAX: 5
  }
};
const comment = {
  rating: -1,
  text: ``
};

const AddReview = (props) => {
  const {film, isValid, onChangeIsValid, postComment} = props;

  const _handleInputChange = (evt) => {
    const target = evt.target;
    const value = target.value;

    if (target.name === `rating`) {
      comment.rating = Number(value);
    }

    if (target.name === `review-text`) {
      comment.text = value;
    }

    const validateStatus = _commentValidate(comment);
    onChangeIsValid(validateStatus);
  };

  const _handlerFormSubmit = (evt) => {
    evt.preventDefault();

    postComment(film.id, comment);

    evt.target.reset();
  };

  const _commentValidate = (filmComment) => {
    const validateErrors = [];

    validateErrors.push(filmComment.text.length < validateRule.TEXT.MIN);
    validateErrors.push(filmComment.text.length > validateRule.TEXT.MAX);
    validateErrors.push(filmComment.rating < validateRule.RANKING.MIN);
    validateErrors.push(filmComment.rating > validateRule.RANKING.MAX);

    return !validateErrors.some((error) => error);
  };

  return (
    <section className="movie-card movie-card--full">
      <div className="movie-card__header">
        <div className="movie-card__bg">
          <img src={film.backgroundImage} alt={film.name} />
        </div>

        <h1 className="visually-hidden">WTW</h1>

        <PageHeader breadcrumbs film={film}/>

        <div className="movie-card__poster movie-card__poster--small">
          <img src={film.posterImage} alt={`${film.name} poster`} width="218" height="327"/>
        </div>
      </div>

      <div className="add-review">
        <form action="#" className="add-review__form" onSubmit={_handlerFormSubmit} onChange={_handleInputChange}>
          <div className="rating">
            <div className="rating__stars">
              {new Array(STARS_COUNT).fill(``).map((_, index) => (
                <Fragment key={`${index}_star`}>
                  <input className="rating__input" id={`star-${index}`} type="radio" name="rating" value={index + 1} />
                  <label className="rating__label" htmlFor={`star-${index}`}>Rating ${index}</label>
                </Fragment>
              ))}
            </div>
          </div>

          <div className="add-review__text">
            <textarea className="add-review__textarea" name="review-text" id="review-text" placeholder="Review text" />
            <div className="add-review__submit">
              {isValid && <button className="add-review__btn" type="submit">Post</button>}
            </div>

          </div>
        </form>
      </div>

    </section>
  );
};

AddReview.propTypes = {
  film: PropTypes.exact({
    id: PropTypes.number,
    name: PropTypes.string,
    posterImage: PropTypes.string,
    previewImage: PropTypes.string,
    backgroundImage: PropTypes.string,
    backgroundColor: PropTypes.string,
    description: PropTypes.string,
    rating: PropTypes.number,
    scoresCount: PropTypes.number,
    director: PropTypes.string,
    starring: PropTypes.arrayOf(PropTypes.string),
    runTime: PropTypes.number,
    genre: PropTypes.string,
    released: PropTypes.number,
    isFavorite: PropTypes.bool,
    videoLink: PropTypes.string,
    previewVideoLink: PropTypes.string,
  }).isRequired,
  onChangeIsValid: PropTypes.func,
  isValid: PropTypes.bool,
  postComment: PropTypes.func,
};

const mapStateToProps = (state, props) => {
  const filmId = props.match.params.id;

  return {

    film: selectors.films.getFilmById(state, filmId),
  };
};

const mapDispatchToProps = (dispatch) => ({
  postComment: (filmId, {rating, text}) => dispatch(operations.comments.storeComment(filmId, rating, text)),
});

export {AddReview};

export default connect(mapStateToProps, mapDispatchToProps)(AddReview);
