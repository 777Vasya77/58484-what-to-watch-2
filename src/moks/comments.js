export const COMMENTS = [
  {
    'id': 1,
    'user': {
      'id': 18,
      'name': `Sophie`
    },
    'rating': 2,
    'comment': `Poised to be an instant classic, almost everything about this film is phenomenal - the acting, the cinematography, the discography, etc.`,
    'date': `2019-10-18T19:05:06.212Z`
  },
  {
    'id': 2,
    'user': {
      'id': 16,
      'name': `Mollie`
    },
    'rating': 2,
    'comment': `This movie really touched my heart, it really is the best movie of the year and everyone should see this masterpiece.`,
    'date': `2019-10-29T19:05:06.212Z`
  },
  {
    'id': 3,
    'user': {
      'id': 14,
      'name': `Corey`
    },
    'rating': 5,
    'comment': `A movie that will take you to another world full of emotions.`,
    'date': `2019-10-02T19:05:06.212Z`
  },
  {
    'id': 4,
    'user': {
      'id': 18,
      'name': `Sophie`
    },
    'rating': 4,
    'comment': `This movie is perfect in all its categories: credits, sound track, production, casting, writing, photography, editing, acting, and direction.\nI was amazed with the freedom of the use of the camera. This movie will change the way movies are made. Slow-mo, stills, black and white, and color were all used to brilliant effect.`,
    'date': `2019-10-27T19:05:06.212Z`
  },
  {
    'id': 5,
    'user': {
      'id': 18,
      'name': `Sophie`
    },
    'rating': 5,
    'comment': `I personally found this movie to be boring. Definitely one of the most boring movies I've ever seen.`,
    'date': `2019-10-20T19:05:06.212Z`
  },
  {
    'id': 6,
    'user': {
      'id': 14,
      'name': `Corey`
    },
    'rating': 5,
    'comment': `I personally found this movie to be boring. Definitely one of the most boring movies I've ever seen.`,
    'date': `2019-10-20T19:05:06.212Z`
  },
  {
    'id': 7,
    'user': {
      'id': 13,
      'name': `Zak`
    },
    'rating': 3,
    'comment': `Unfortunately we don't have a reliable way to tell the true ratings of a movie.`,
    'date': `2019-10-26T19:05:06.212Z`
  },

];
