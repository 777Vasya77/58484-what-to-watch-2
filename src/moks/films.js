export const FILMS = [
  {
    title: `Fantastic Beasts: The Crimes of Grindelwald`,
    image: `fantastic-beasts-the-crimes-of-grindelwald.jpg`
  },
  {
    title: `Aviator`,
    image: `aviator.jpg`
  },
  {
    title: `Bohemian Rhapsody`,
    image: `bohemian-rhapsody.jpg`
  },
  {
    title: `Macbeth`,
    image: `macbeth.jpg`
  },
  {
    title: `Mindhunter`,
    image: `mindhunter.jpg`
  },
  {
    title: `War of the Worlds`,
    image: `war-of-the-worlds.jpg`
  },
  {
    title: `Johnny English`,
    image: `johnny-english.jpg`
  },
  {
    title: `We Need to Talk about Kevin`,
    image: `we-need-to-talk-about-kevin.jpg`
  },
];
