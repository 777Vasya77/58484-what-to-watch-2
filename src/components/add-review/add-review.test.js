import React from 'react';
import renderer from 'react-test-renderer';
import {AddReview} from "~/components/add-review/add-review";
import Film from '~/models/film';
import {FILMS} from '~/moks/test-moks';

jest.mock(`~/components/page-header/page-header`, () => `page-header`);
const film = new Film(FILMS[0]);

describe(`AddReview component tests`, () => {
  it(`Component render correctly`, () => {
    const tree = renderer
        .create(
            <AddReview
              film={film}
              history={({push: () => {}})}
            />
        ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
