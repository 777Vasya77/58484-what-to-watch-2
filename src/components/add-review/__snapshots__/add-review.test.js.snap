// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddReview component tests Component render correctly 1`] = `
<section
  className="movie-card movie-card--full"
>
  <div
    className="movie-card__header"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="Macbeth"
        src="https://es31-server.appspot.com/wtw/static/film/background/Macbeth.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <page-header
      breadcrumbs={true}
      film={
        Film {
          "backgroundColor": "#F1E9CE",
          "backgroundImage": "https://es31-server.appspot.com/wtw/static/film/background/Macbeth.jpg",
          "description": "Macbeth, the Thane of Glamis, receives a prophecy from a trio of witches that one day he will become King of Scotland. Consumed by ambition and spurred to action by his wife, Macbeth murders his king and takes the throne for himself.",
          "director": "Justin Kurzel",
          "genre": "Drama",
          "id": 1,
          "isFavorite": false,
          "name": "Macbeth",
          "posterImage": "https://es31-server.appspot.com/wtw/static/film/poster/Macbeth.jpg",
          "previewImage": "https://es31-server.appspot.com/wtw/static/film/preview/macbeth.jpg",
          "previewVideoLink": "https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4",
          "rating": 3.3,
          "released": 2015,
          "runTime": 113,
          "scoresCount": 48798,
          "starring": Array [
            "Michael Fassbender",
            "Marion Cotillard",
            "Jack Madigan",
          ],
          "videoLink": "http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4",
        }
      }
    />
    <div
      className="movie-card__poster movie-card__poster--small"
    >
      <img
        alt="Macbeth poster"
        height="327"
        src="https://es31-server.appspot.com/wtw/static/film/poster/Macbeth.jpg"
        width="218"
      />
    </div>
  </div>
  <div
    className="add-review"
  >
    <form
      action="#"
      className="add-review__form"
      onChange={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="rating"
      >
        <div
          className="rating__stars"
        >
          <input
            className="rating__input"
            id="star-0"
            name="rating"
            type="radio"
            value={1}
          />
          <label
            className="rating__label"
            htmlFor="star-0"
          >
            Rating $
            0
          </label>
          <input
            className="rating__input"
            id="star-1"
            name="rating"
            type="radio"
            value={2}
          />
          <label
            className="rating__label"
            htmlFor="star-1"
          >
            Rating $
            1
          </label>
          <input
            className="rating__input"
            id="star-2"
            name="rating"
            type="radio"
            value={3}
          />
          <label
            className="rating__label"
            htmlFor="star-2"
          >
            Rating $
            2
          </label>
          <input
            className="rating__input"
            id="star-3"
            name="rating"
            type="radio"
            value={4}
          />
          <label
            className="rating__label"
            htmlFor="star-3"
          >
            Rating $
            3
          </label>
          <input
            className="rating__input"
            id="star-4"
            name="rating"
            type="radio"
            value={5}
          />
          <label
            className="rating__label"
            htmlFor="star-4"
          >
            Rating $
            4
          </label>
        </div>
      </div>
      <div
        className="add-review__text"
      >
        <textarea
          className="add-review__textarea"
          id="review-text"
          name="review-text"
          placeholder="Review text"
        />
        <div
          className="add-review__submit"
        />
      </div>
    </form>
  </div>
</section>
`;
